### PR TITLE
feat: Record AWS Lambda coldstarts

### PR DESCRIPTION
**Pull request by serkan-ozal**
## Which problem is this PR solving?

Adds recording AWS Lambda invocation coldstarts as Lambda invocation span attributes.

## Short description of the changes

Decides whether an AWS Lambda invocation is coldstart or not by taking care of the following cases:
- Non-first invocations are warm, not coldstart
- If sandbox environment is initialized with provisioned concurrency, every invocations are warm, not coldstart
- If sandbox has been initialized [proactively](https://aaronstuyvenberg.com/posts/understanding-proactive-initialization) before the actual request, every invocations are warm, not coldstart

### CODE REVIEW

**Review by tylerbenson — APPROVED** [review]

File: plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts [hunk 1]
```diff
@@ -164,16 +174,47 @@ export class AwsLambdaInstrumentation extends InstrumentationBase<AwsLambdaInstr
...
+        if (
+          process.env.AWS_LAMBDA_INITIALIZATION_TYPE ===
+          'provisioned-concurrency'
+        ) {
+          // If sandbox environment is initialized with provisioned concurrency,
+          // even the first requests should not be considered as coldstart.
+          requestIsColdStart = false;
+        } else {
+          // Check whether it is proactive initialization or not:
+          // https://aaronstuyvenberg.com/posts/understanding-proactive-initialization
```
Comment (pragmaticivan) [inline plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts:204]: 🤯

**Review by pragmaticivan — APPROVED** [review]

**Review by tylerbenson — APPROVED** [review]

File: plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts [hunk 2]
```diff
@@ -44,6 +44,7 @@ import {
   AWSXRayPropagator,
 } from '@opentelemetry/propagator-aws-xray';
 import {
+  SEMATTRS_FAAS_COLDSTART,
```
Comment (dyladan) [inline plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts:47]: Nit: Would be nice to update the semconv package and use the new constant. Not blocking

**Review by dyladan — APPROVED** [review]
Looks reasonable to me. I think this can be merged. I left a minor suggestion, but if you don't want to do it that's fine. Just comment and we'll merge this when you're ready.

**Comment by serkan-ozal** [comment]
> Looks reasonable to me. I think this can be merged. I left a minor suggestion, but if you don't want to do it that's fine. Just comment and we'll merge this when you're ready.

@dyladan Are you suggesting to update semconv package and update all semconv attributes in all places? If so, I think it should be done in a separate PR. I have used `SEMATTRS_...` properties, because as far as I see, it is the general approach  by other instrumentations too.

**Comment by dyladan** [comment]
> > Looks reasonable to me. I think this can be merged. I left a minor suggestion, but if you don't want to do it that's fine. Just comment and we'll merge this when you're ready.
> 
> @dyladan Are you suggesting to update semconv package and update all semconv attributes in all places accordingly? If so, I think it should be done in a separate PR. I have used `SEMATTRS_...` properties, because as far as I see, it is the general approach used by other instrumentation plugins too.

I had been thinking only the newly introduced attribute would use the new format. The old formats are still exported for backward compatibility. but leaving it for a followup is also fine.

File: plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts (hunk 2 above)
Comment (serkan-ozal) [reply plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts]: @dyladan ok, done. just used that new attribute (`faas.coldstart`) from new constant (experimental attribute through incubating)

**Pull request merged**